core/src/main/java/org/elasticsearch/index/translog/Checkpoint.java

```diff
@@ -62,66 +67,78 @@
         + Long.BYTES // generation
         + CodecUtil.footerLength();
 
-    static final int LEGACY_NON_CHECKSUMMED_FILE_LENGTH = Integer.BYTES  // ops
-            + Long.BYTES // offset
-            + Long.BYTES; // generation
-
-    Checkpoint(long offset, int numOps, long generation, long globalCheckpoint) {
+    /**
+     * Create a new translog checkpoint.
+     *
+     * @param offset           the current offset in the translog
+     * @param numOps           the current number of operations in the translog
+     * @param generation       the current translog generation
+     * @param minSeqNo         the current minimum sequence number of all operations in the translog
+     * @param maxSeqNo         the current maximum sequence number of all operations in the translog
+     * @param globalCheckpoint the last-known global checkpoint
+     */
+    Checkpoint(long offset, int numOps, long generation, long minSeqNo, long maxSeqNo, long globalCheckpoint) {
+        assert minSeqNo <= maxSeqNo;
         this.offset = offset;
         this.numOps = numOps;
         this.generation = generation;
+        this.minSeqNo = minSeqNo;
+        this.maxSeqNo = maxSeqNo;
         this.globalCheckpoint = globalCheckpoint;
     }
 
     private void write(DataOutput out) throws IOException {
         out.writeLong(offset);
         out.writeInt(numOps);
         out.writeLong(generation);
+        out.writeLong(minSeqNo);
+        out.writeLong(maxSeqNo);
         out.writeLong(globalCheckpoint);
     }
 
-    static Checkpoint readChecksummedV2(DataInput in) throws IOException {
-        return new Checkpoint(in.readLong(), in.readInt(), in.readLong(), in.readLong());
+    static Checkpoint emptyTranslogCheckpoint(final long offset, final long generation, final long globalCheckpoint) {
+        final long minSeqNo = SequenceNumbersService.NO_OPS_PERFORMED;
+        final long maxSeqNo = SequenceNumbersService.NO_OPS_PERFORMED;
+        return new Checkpoint(offset, 0, generation, minSeqNo, maxSeqNo, globalCheckpoint);
     }
 
-    // reads a checksummed checkpoint introduced in ES 5.0.0
-    static Checkpoint readChecksummedV1(DataInput in) throws IOException {
-        return new Checkpoint(in.readLong(), in.readInt(), in.readLong(), SequenceNumbersService.UNASSIGNED_SEQ_NO);
+    static Checkpoint readChecksummedV2(final DataInput in) throws IOException {
+        return new Checkpoint(in.readLong(), in.readInt(), in.readLong(), in.readLong(), in.readLong(), in.readLong());
     }
 
-    // reads checkpoint from ES < 5.0.0
-    static Checkpoint readNonChecksummed(DataInput in) throws IOException {
-        return new Checkpoint(in.readLong(), in.readInt(), in.readLong(), SequenceNumbersService.UNASSIGNED_SEQ_NO);
+    // reads a checksummed checkpoint introduced in ES 5.0.0
+    static Checkpoint readChecksummedV1(final DataInput in) throws IOException {
+        final long minSeqNo = SequenceNumbersService.NO_OPS_PERFORMED;
```

Comment:
copying the commit comment as the code moved around:

> these should be UNASSIGNED_SEQ_NO no? (unless maybe if total count is 0, but I'm not sure it's worth the code song and dance it will require)